llvm/lib/Transforms/IPO/FunctionAttrs.cpp

```diff
@@ -118,16 +118,31 @@ static void addLocAccess(MemoryEffects &ME, const MemoryLocation &Loc,
                          ModRefInfo MR, AAResults &AAR) {
   // Ignore accesses to known-invariant or local memory.
   MR &= AAR.getModRefInfoMask(Loc, /*IgnoreLocal=*/true);
-  if (isNoModRef(MR))
+  if (isNoModRef(MR)) {
+    // Mask out errno, can never alias this known-invariant memory location.
+    ME -= ME.errnoMemOnly();
     return;
+  }
 
   const Value *UO = getUnderlyingObjectAggressive(Loc.Ptr);
-  if (isa<AllocaInst>(UO))
+  if (isa<AllocaInst>(UO)) {
+    ME -= ME.errnoMemOnly();
```

Comment:
I don't understand what is going on here. You're saying that any function that accesses an alloca doesn't accesses errno?

Generally, it doesn't make sense to mask out something from ME after the fact.

Reply:
I feel like the original intent was to try to remove errno, if, say, an argument ptr is an alloca for that memory location; which is however not what is happening here. Looking at this better again, I think there's not much we can reliably infer here, and, once again, it should be up to getModRefInfo to determine whether the underlying ptr is alloca memory? Updated this, meanwhile.